File: src/main/java/de/tuberlin/spreadalgo/SpreadedMarkers.java
package de.tuberlin.spreadalgo;

import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedList;

/**
 * Created by faehndrich on 06.05.15.
 */
public class SpreadedMarkers {
    HashMap<Node, Collection<SpreadingStep>> markers = new HashMap<Node, Collection<SpreadingStep>>();

    public void addAll(Collection<SpreadingStep> data) {
        for(SpreadingStep step : data) {
            Node target = step.getTargetNode();
            getInputForTarget(target).add(step);
        }
    }

    public Collection<Node> getTargetNodes() {
        return markers.keySet();
    }

    public Collection<SpreadingStep> getInputForTarget(Node targetNode) {
        Collection<SpreadingStep> ret = markers.get(targetNode);
        if(ret == null) {
            ret = new LinkedList<SpreadingStep>();
            markers.put(targetNode,ret);
        }
        return ret;
    }
}

File: marker_passing/spreaded_markers.py
from __future__ import annotations
from collections import defaultdict
from typing import Collection, Dict, List, TYPE_CHECKING

if TYPE_CHECKING:
    from .node import Node
    from .spreading_step import SpreadingStep


class SpreadedMarkers:
    """
    Collects spreading steps grouped by their target node, ready for in-function processing.
    """

    def __init__(self) -> None:
        self._markers: Dict["Node", List["SpreadingStep"]] = defaultdict(list)

    def add_all(self, steps: Collection["SpreadingStep"]) -> None:
        for step in steps:
            target = step.get_target_node()
            self._markers[target].append(step)

    def get_target_nodes(self) -> Collection["Node"]:
        return self._markers.keys()

    def get_input_for_target(self, target_node: "Node") -> List["SpreadingStep"]:
        return self._markers[target_node]
